[FLINK-27693][state] Support local recovery for non-materialized part

fredia opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Support local recovery for the non-materialized part of changelog.


## Brief change log

- [8cc0d2](https://github.com/apache/flink/commit/8cc0d210c109768c705f1e4d774cf577e0d56f12) add localConfig param
- [8f7f7dd](https://github.com/apache/flink/commit/8f7f7dd21d5b652f3de5bca0858510ca0b0d8d66) write to double stream
- [ed4600](https://github.com/apache/flink/commit/ed46006ecf037d807be54dbf5725b5dc55f0c23c) add local state registry to discard local changelog handle
- [8088f0](https://github.com/apache/flink/commit/8088f02c705971404a038b783e6a8c2093cf43dd) return local handle after snapshot


## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added tests and can be verified as follows:
- `ChangelogLocalRecoveryITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/StateChangeFsUploader.java:
```diff
@@ -19,50 +19,30 @@
  * {@link FileSystem}.
  */
-public class StateChangeFsUploader implements StateChangeUploader {
+public class StateChangeFsUploader extends AbstractStateChangeFsUploader {
```
`AbstractStateChangeFsUploader` was extracted and is extended by `DuplicatingStateChangeFsUploader` and `StateChangeFsUploader`.
It has some `protected` methods that are used by descendants; and no `abstract` methods. And the calling methods in descendants (`upload`) have some duplication.

That means there is something wrong with the design.
How about applying template method pattern and having `upload()` in the base class and some `abstract` methods to customize it (i.e. create stream and handle errors)?

(alternatively, the two classes can be just implementers; code can be shared via composition or just static methods)

WDYT?

fredia replied:
Nice suggestion, I extracted the logic of creating stream into `prepareStream()`.

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -329,14 +343,28 @@ public void confirm(SequenceNumber from, SequenceNumber to) {
+
+            // transfer the control of localHandle to localStateRegistry.
+            for (UploadResult result : uploaded.subMap(from, to).values()) {
+                changelogRegistry.stopTracking(result.localStreamHandle);
+                localStateRegistry.register(result.localStreamHandle, checkpointId);
```
Null check for `result.localStreamHandle`?

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -329,14 +343,28 @@ public void confirm(SequenceNumber from, SequenceNumber to) {
+                changelogRegistry.stopTracking(result.localStreamHandle);
+                localStateRegistry.register(result.localStreamHandle, checkpointId);
+            }
+            localStateRegistry.unRegister(checkpointId);
```
Just to clarify: we only retain local state for the last checkpoint (num retained checkpoint is ignored). Same as the current local recovery, right?

fredia replied:
👍Right, we only retain local state for the last one checkpoint.

rkhachatryan replied:
Thanks!

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadResult.java:
```diff
@@ -21,12 +21,15 @@
 @Internal
 public final class UploadResult {
     public final StreamStateHandle streamStateHandle;
+    public final StreamStateHandle localStreamHandle;
```
Mark `@Nullable`?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendLocalHandle.java:
```diff
@@ -0,0 +1,116 @@
+        return remoteHandle.getStateHandleId();
+    }
+
+    @Override
+    public void discardState() throws Exception {}
```
:+1: IIUC, the discard logic is correct:
1. `(Changelog)TaskLocalStateStore` calls this no-op method
1. `LocalStateRegistry` calls `discardState()` on "low-level" handle (`FileStateHandle`) - on checkpoint confirmation
1. `TaskChangelogRegistry` - calls `discardState()` on "low-level" handle (`FileStateHandle`) - when upload is no more needed
1. As opposed to "normal" local state, local changelog state is *not* discarded on recovery (because of 1); but it will be discarded on the 1st checkpoint confirmation

Can you confirm that?

Although I found it tricky to understand, so it makes sense to document it somewhere, WDYT?

(and also to cover with tests).

rkhachatryan replied:
Probably this case is not handled: the last checkpoint(s) aborted and then job terminated?

fredia replied:
> 1. (Changelog)TaskLocalStateStore calls this no-op method

Yes, `TaskLocalStateStore` call no-op method, but `TaskLocalStateStore` would delete the whole chk folder(`changelog_chk_` and `chk_`) when confirm checkpoint. 

> 4. As opposed to "normal" local state, local changelog state is not discarded on recovery (because of 1); but it will be discarded on the 1st checkpoint confirmation

Like "normal" local state, local changelog state would **be discarded** on recovery, it would be delete when `deleteDirectory()`, because of `recursive=true`.

```java
protected void deleteDirectory(File directory) throws IOException {
        Path path = new Path(directory.toURI());
        FileSystem fileSystem = path.getFileSystem();
        if (fileSystem.exists(path)) {
            fileSystem.delete(path, true);
        }
    }
```

The total discard logic of local dstl files is:
1. Register files to `TaskChangelogRegistry#startTracking()` on `StateChangeUploader#upload()`.
2.  Pass control of the file to `LocalStateRegistry#register()` **when** `confirm()` , files of the previous checkpoint will be unregistered/deleted by `LocalStateRegistry#unRegister()`  at the same time ; for remote file, we pass control of the file to JM.
3. When `materialized()` or `checkpointSubsumed()` is called, `TaskChangelogRegistry#notUse()` is responsible for deleting local files. BTW, it is possible that the file has already been truncated by `LocalStateRegistry`.

rkhachatryan replied:
> Yes, TaskLocalStateStore call no-op method, but TaskLocalStateStore would delete the whole chk folder(changelog_chk_ and chk_) when confirm checkpoint.

I was assuming that deleting the whole directory still does not guarantee deleting the changelog (because it can reside outside); that's why there is an explicit call `taskStateSnapshot.discardState();` in "normal" local state store.
But either way, recursive deletion or the next checkpoint should discard it.

> The total discard logic of local dstl files is

Thanks for the explanation. 
Could you add this comments ("total discard logic") somewhere int the code?

And I still don't see how this case is handled:
> the last checkpoint(s) aborted and then job terminated?

fredia replied:
> I was assuming that deleting the whole directory still does not guarantee deleting the changelog (because it can reside outside); that's why there is an explicit call taskStateSnapshot.discardState(); in "normal" local state store.
But either way, recursive deletion or the next checkpoint should discard it.

Yes, deleting the whole directory still does not guarantee deleting the changelog (because it can reside outside). `ChangelogStateHandleStreamImpl` are placed in `/dstl`. 
For **Local** `ChangelogStateHandleStreamImpl`, they are placed in <LOCAL_DIR>/taskowned, and they would be discarded by  `TaskChangelogRegistry` and  `LocalStateRegistry`. 
So,  *recursive deletion `changelog_chk_` directory* + *`LocalStateRegistry`* + *`TaskChangelogRegistry`* are used to guarantee non-materialized part's deletion.

> Could you add this comments ("total discard logic") somewhere in the code?

Sure, I will add this comment in `ChangelogTaskLocalStateStore`.

> Probably this case is not handled: the last checkpoint(s) aborted and then job terminated?

You are right, "the last checkpoint(s) aborted and then job terminated"  can't be handled in this code. 
Regardless of the deletion of local dstl, do you know how **remote dstl** handles this case?  if one `StateChangeSet` is uploaded to `/dstl`,  will it be removed on abort?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,115 @@
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** This registry manages handles which is written for local recovery. */
+public class LocalStateRegistry {
```
I'm thinking about the case when local changelog state explodes and causes `no space left on device`.
The host might become unresponsive and it will be difficult to debug such an issue.

To ease the debug, we should probably monitor the space taken by the changelog; and stop writing locally or throw an exception when it takes too much (above some configured limit).
This class seems the best place for such monitoring.

This can be a separate ticket though, at least if local recovery is not the default.

WDYT?

fredia replied:
Nice suggestion, another tick for monitoring the space taken by the changelog could be created later, but we may not have enough time to finish it in v1.16(Changelog V2).

rkhachatryan reviewed (commented):
Thanks a lot for the PR @fredia
I don't see any major issues with it.

I've left some comments, PTAL (probably after rebasing).

fredia commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingOutputStreamWithPos.java:
```diff
@@ -0,0 +1,94 @@
+    @Override
+    public void write(int b) throws IOException {
+        outputStream.write(b);
+        try {
+            secondaryStream.write(b);
+        } catch (Exception ex) {
+            LOG.warn("Exception encountered during write to secondary stream");
+        }
```
I modified it with reference to the method in `DuplicatingCheckpointOutputStream`.

fredia commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingOutputStreamWithPos.java:
```diff
@@ -0,0 +1,94 @@
+ * wraps a primary and a secondary OutputStream and duplicates all writes into both streams. The
+ * difference is that this stream does not delete the file when {@link #close()}.
+ */
+class DuplicatingOutputStreamWithPos extends OutputStreamWithPos {
```
We have three options to wrap a output stream：

 1. Path ->`FSDataOutputStream` -> (compress) -> `OutputStream` -> (buffer) -> `BufferedOutputStream` -> (add pos) `OutputStreamWithPos or DuplicatingOutputStreamWithPos`.  
 2. Path -> `CheckpointStateOutputStream or DuplicatingCheckpointOutputStream ` -> (compress) ->  `OutputStream`  -> (counting written bytes?).
 3. Path -> `FSDataOutputStream` -> (compress) -> `OutputStream` -> (reuse `CheckpointStateOutputStream` if it offers a constructor with `OutputStream` parameter) -> `CheckpointStateOutputStream or DuplicatingCheckpointOutputStream`.
 
Option 1 is the current implementation.

Is option 2 what you want to express?  I think the `getPos()` before and after compression are not the same, maybe we still need a wrapper class to count written bytes.

Option 3 needs to modify the constructor method of `CheckpointStateOutputStream`.

fredia replied:
Because of [FLINK-28602](https://issues.apache.org/jira/browse/FLINK-28602), we also should override `close()` method.

rkhachatryan replied:
Yes, I meant something like (2):
- `FileBasedStateOutputStream` 
- decorated by `StreamCompressionDecorator`
- wrapped by `BufferedOutputStream` 
- wrapped into `CheckpointStateOutputStream` 

Adding buffering and `getPos` might be problematic though. 

I just noticed that the current local recovery stream (for Heap and RocksDB metadata) is also NOT buffered. 
So I'd add a ticket to add buffering there, and then reconsider this issue (leaving the code as is until then).

WDYT?

fredia replied:
Sounds good, thanks.

fredia commented:
Hi @rkhachatryan, I updated this PR to handle "the last checkpoint(s) aborted and then job terminated" case, and added `LocalStateRegistryTest`. Could you please take a look again?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,117 @@
+    public StreamStateHandle register(StreamStateHandle handle, long checkpointID) {
+        synchronized (registeredHandles) {
+            checkState(open, "Attempt to register state to closed LocalStateRegistry.");
+            if (registeredHandles.containsKey(handle)) {
+                long pre = registeredHandles.get(handle);
+                if (checkpointID > pre) {
+                    registeredHandles.put(handle, checkpointID);
+                }
+            } else {
+                registeredHandles.put(handle, checkpointID);
+            }
```
How about simplifying it with:
```
registeredHandles.put(
    handle,
    Math.max(checkpointID, registeredHandles.getOrDefault(handle, -1L)));
```
?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,117 @@
+                registeredHandles.put(handle, checkpointID);
+            }
+        }
+        return handle;
```
The returned value is never used.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,117 @@
+        return handle;
+    }
+
+    public void unRegister(long upTo) {
```
I'd add some javadoc explaining when this method is called.

I'd also rename it to reflect what it actually does. WDYT about `discardUpToCheckpoint(long checkpointID)`?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,117 @@
+                    registeredHandles.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Entry<StreamStateHandle, Long> entry = iterator.next();
+                if (entry.getValue() < upTo) {
+                    handles.add(entry.getKey());
+                    iterator.remove();
+                }
```
I'm afraid this is not enough because the actual underlying file might be used by multiple state backends / tasks (`AbstractStateChangeFsUploader` is shared by multiple of them).

So we either need to 
- track usage by backend (similar to `TaskChangelogRegistry`)
- or create a separate duplicating stream per state backend - and manage separate offsets

fredia replied:
Yes,  the actual underlying file might be used by multiple state backends / tasks,  and I think the file has been **implicitly** tracked by every backend:

`LocalStateRegistry#register()` is called when `ChangelogKeyedStateBackend#notifyCheckpointComplete` ->`FsStateChangelogWriter#confirm()`.  When a file is still referenced by a state backend,  the state backend would update its `LastUsedCheckpointID`. Suppose there are three state backend data in file 1,  
```
File 1:
    backend A : c1, c2
    backend B : c1
    backend C : c2, c3
```

- checkpoint 1 confirm, LastUsedCheckpointID = 1,  upTo=1 , won't delete;
- checkpoint 2 confirm, LastUsedCheckpointID = 2, upTo=2, won't delete;
- checkpoint 3 confirm, LastUsedCheckpointID = 3, upTo=3, won't delete;
- checkpoint 4 confirm,  LastUsedCheckpointID = 3, upTo=4, **delete**;


I think the role of `LocalStateRegistry` is similar to `SharedStateRegisrty` on JM. when passing the control of remote handles to JM, the remote handles are not explicitly tracked by multi-state backends, so I don't think `LocalStateRegistry` needs to be either.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public void close() {
+        synchronized (registeredHandles) {
+            open = false;
+        }
+    }
```
As mentioned above, this method should probably discard the remaining state.

If not, and if this class will stay per task (likely not); I'd remove `open` because it requires adding `syncrhonized` sections, which IMO doesn't worth enforcing the lifecycle without any cleanup.

fredia replied:
Yes, it doesn't worth enforcing the lifecycle without any cleanup, removed it.

rkhachatryan reviewed (changes requested):
Thanks for updating the PR @fredia 
I've left some comments, PTAL.

The biggest concern is sharing of local changelog files by multiple tasks.

fredia commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ChangelogTaskLocalStateStore.java:
```diff
@@ -181,6 +226,16 @@ public void pruneCheckpoints(LongPredicate pruningChecker, boolean breakOnceChec
     @Override
     public CompletableFuture<Void> dispose() {
         deleteMaterialization(id -> true);
+        // delete all ChangelogStateHandle in taskowned directory.
+        discardExecutor.execute(
+                () ->
+                        syncDiscardFileForCollection(
```
Yes, this will delete other job's states.
I moved `aid_xx/taskowned` folder to `aid_xxx/jid_xxx/taskowned` to avoid this.

> Why don't we put this logic in LocalStateRegistry - that would allow to avoid clashing and guessing the locations

Because the `taskowned` folder needs to be deleted when `dispose()`, and `LocalStateRegistry` can only delete the files in the folder, can't delete the folder itself.

fredia commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManager.java:
```diff
@@ -101,6 +103,7 @@ public TaskExecutorLocalStateStoresManager(
         this.discardExecutor = discardExecutor;
         this.lock = new Object();
         this.closed = false;
+        this.localStateRegistry = new LocalStateRegistry(discardExecutor);
```
The lifecycle of `LocalStateRegistry` can only be per TM, because `LocalStateRegistry` is running on TM, it can't access other tasks on other TMs.

rkhachatryan replied:
Yes, I meant per job per TM, so that state backends of the same job running on the same TM should share this object.

rkhachatryan replied:
Yes, I meant per job per TM, so that state backends of the same job running on the same TM should share this object.

fredia commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,117 @@
+     * All registered handles. (handle,checkpointID) represents a handle and the latest checkpoint
+     * that refer to this handle.
+     */
+    private final Map<StreamStateHandle, Long> registeredHandles;
```
1. Because `StreamStateHandle.discardState()` needs to be called when `unRegister`, `PhysicalStateHandleID` doesn't have access to  `discardState()`;
2. 👍Renamed `registeredHandles` to `handleToLastUsedCheckpointID`.

rkhachatryan replied:
1. How about `Map<PhysicalStateHandleID, Tuple<StreamStateHandle, Long>>`?

fredia commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java:
```diff
@@ -0,0 +1,117 @@
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** This registry manages handles which is written for local recovery. */
+public class LocalStateRegistry implements Closeable {
```
> I guess it's discarding of local changelog; which can't be discarded by ChangelogTaskLocalStateStore because they are not bound to a specific checkpoint.
1. Yes, I updated the javadoc of `LocalStateRegistry`;
2. Renamed `LocalStateRegistry` to `LocalChangelogRegistry`.
3. It has to be used by `flink-dstl` module, I annotated it with `@Internal`

fredia commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ChangelogTaskLocalStateStore.java:
```diff
@@ -97,6 +101,47 @@ private void updateReference(long checkpointId, TaskStateSnapshot localState) {
+        //   confirm cp3, do nothing
+        //   -> if failover, restore from local cp3 will fail, because m1 does not exist, c1 may not
+        // exist either(depend on BatchingStateChangeUploadScheduler).
+        File[] fileInTaskOwned =
+                new File(getLocalTaskOwnedDirectory(getLocalRecoveryDirectoryProvider()).toUri())
+                        .listFiles();
+        syncDiscardFileForCollection(
+                fileInTaskOwned == null ? Collections.emptyList() : Arrays.asList(fileInTaskOwned));
```
You're right, I moved the deletion to `LocalStateRegistry#prune` to handle "the last checkpoint(s) aborted and then job terminated".

fredia commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ChangelogTaskLocalStateStore.java:
```diff
@@ -97,6 +101,47 @@ private void updateReference(long checkpointId, TaskStateSnapshot localState) {
     }
 
+    public static Path getLocalTaskOwnedDirectory(LocalRecoveryDirectoryProvider provider) {
+        File outDir = provider.selectAllocationBaseDirectory(0);
```
Yes, it should be rotated.
I updated it with reference to :

```
public File allocationBaseDirectory(long checkpointId) {
        return selectAllocationBaseDirectory(
                (((int) checkpointId) & Integer.MAX_VALUE) % allocationBaseDirs.length);
    }
```

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -346,12 +376,42 @@ private static ChangelogStateHandleStreamImpl buildHandle(
+            List<Tuple2<StreamStateHandle, Long>> localTuples = new ArrayList<>();
+            for (UploadResult uploadResult : results.values()) {
+                if (uploadResult.getLocalStreamHandleStateHandle() != null) {
+                    localTuples.add(
+                            Tuple2.of(
+                                    uploadResult.getLocalStreamHandleStateHandle(),
+                                    uploadResult.getOffset()));
```
The local offset is now passed and saved in `uploadResult`, but not actually used here.

fredia commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorage.java:
```diff
@@ -104,17 +126,24 @@ public FsStateChangelogStorage(
+        this.localRecoveryConfig = localRecoveryConfig;
+        if (localRecoveryConfig.isLocalRecoveryEnabled()) {
+            this.localChangelogRegistry =
+                    new LocalChangelogRegistry(Executors.newSingleThreadExecutor());
```
This and `defaultChangelogRegistry` would create two executors for each job, how about passing the `TaskManagerServices.ioExecutor` for this?

rkhachatryan replied:
`TaskManagerServices.ioExecutor` is used for other jobs as well; re-using it might break resource isolation.
It would make sense to share the threads between `TaskChangelogRegistry` and `LocalChangelogRegistry`, but creating two threads per job on TM seems fine too.

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorage.java:
```diff
@@ -63,6 +66,9 @@ public class FsStateChangelogStorage extends FsStateChangelogStorageForRecovery
+    @Nullable private LocalChangelogRegistry localChangelogRegistry = null;
+
     /** The configuration for local recovery. */
     @Nonnull private final LocalRecoveryConfig localRecoveryConfig;
```
nit: I'd replace this pair `Nullable registry` and `Nonnull config` with some no-op `LocalChangelogRegistry`.
In `FsStateChangelogWriter.confirm()`, we could just check `if (result.localStreamHandle != null) { ... } `

rkhachatryan commented:
Merged manually as 1f9632a07199854c0225bd7f416c038fbf59abe0..52eb7e76b5d66ff5c4d9d4af8a213b5b8f9f8322
(build [succeeded](https://dev.azure.com/khachatryanroman/flink/_build/results?buildId=1600&view=results) after rebasing).
